File: src/services/CareerService.ts
import axios from "axios";
import {RootPath} from "./RootPath";
import Career from "../models/actor/player/Career";

export default class careerService {
    static async createCareer(name: string): Promise<Career> {
        return await (await axios.post(RootPath.Career + name)).data;
    }

    static async getCareer(name: string): Promise<Career> {
        return await (await axios.get(RootPath.Career + name)).data;
    }

    static async updateCareer(name: string, career: Career): Promise<Career> {
        return await (await axios.put(RootPath.Career + name, career)).data;
    }

    static async getCareers(): Promise<Career[]> {
        return await (await axios.get(RootPath.Career)).data;
    }
}

File: server/src/main/java/com/github/genraven/genesys/service/CareerService.java
package com.github.genraven.genesys.service;

import com.github.genraven.genesys.domain.actor.player.Career;
import com.github.genraven.genesys.domain.skill.Skill;
import com.github.genraven.genesys.repository.CareerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CareerService {
    private final CareerRepository careerRepository;
    private final SkillService skillService;

    public Flux<Career> getAllCareers() {
        return careerRepository.findAll();
    }

    public Mono<Career> getCareer(final String name) {
        return careerRepository.findById(name);
    }

    public Mono<Career> createCareer(final String name) {
        return skillService.getSkillsForCurrentCampaign()
                .flatMap(skills -> {
                    final List<Skill> careerSkills = skills.stream().limit(8).toList();
                    final Career career = new Career(name);
                    career.setSkills(careerSkills);
                    return careerRepository.save(career);
                });
    }

    public Mono<Career> updateCareer(final String name, final Career career) {
        return getCareer(name).map(car -> {
            car.setSkills(career.getSkills());
            return car;
        }).flatMap(careerRepository::save);
    }
}
